crates/core/src/operations/optimize.rs

```diff
@@ -1066,6 +1075,29 @@ async fn build_zorder_plan(
     Ok((operation, metrics))
 }
 
+async fn get_file_decryption_properties(
+    file_format_options: &FileFormatRef,
+    file_path: &object_store::path::Path,
+) -> Result<Option<FileDecryptionProperties>, DataFusionError> {
+    let parquet_options = file_format_options.table_options().parquet;
+    if let Some(props) = &parquet_options.crypto.file_decryption {
+        return Ok(Some(props.clone().into()));
+    }
+    if let Some(factory_id) = &parquet_options.crypto.factory_id {
+        // Create a temporary DataFusion session to access the encryption factory
```

Comment:
This is a bit convoluted too

Reply:
Yeah, agree this is convoluted and it seems we have to do it in a couple places. Maybe encapsulate this to a function in file_format_options.rs and eventually find a better way?